File: app/models/status.rb
class Status < ActiveRecord::Base
  belongs_to :project

  validates :title, presence: true
  validate :finish_at_is_after_start_at

  before_validation :set_start_at_to_now_when_nil
  before_validation :set_finish_at_to_nil_when_equal_to_start_at

  def tag
    project.tags.find tag_id
  end

  def tag= tag
    self.tag_id = tag.try :id
  end

  def start_at= time
    super convert_fixnum_to_time(time)
  end

  def finish_at= time
    super convert_fixnum_to_time(time)
  end

private

  def convert_fixnum_to_time time
    time = Time.at time / 1000 if time.is_a? Fixnum
    time
  end

  def set_start_at_to_now_when_nil
    self.start_at = Time.current unless start_at
  end

  def set_finish_at_to_nil_when_equal_to_start_at
    self.finish_at = nil if start_at == finish_at
  end

  def finish_at_is_after_start_at 
    if finish_at && finish_at < start_at
      errors.add :finish_at, :too_short
    end
  end
end

File: app/assets/javascripts/models/status.js
App.Status = DS.Model.extend({
  title: DS.attr(),
  description: DS.attr(),
  startAt: DS.attr('date'),
  finishAt: DS.attr('date'),
  tagId: DS.attr(),

  project: DS.belongsTo('project'),

  tag: function(key, value) {
    var tagId, tagList;

    if (arguments.length === 2) {
      tagId = value ? value.get('id') : null;
      this.set('tagId', tagId);
    } else {
      tagId = this.get('tagId');
      tagList = this.get('project.tagList');
      value = tagList.findProperty('id', tagId);
    }

    return value;
  }.property('tagId', 'project.tagList.[]'),

  tagClassName: function() {
    var tagId = this.get('tagId');
    return tagId ? 'tag-' + tagId : 'no-tag';
  }.property('tagId')
});
